### DISCUSSION TITLE
Динамическая генерация QR кода для печати

### DISCUSSION THREAD

**Discussion by Qb1cc** [totumonline/totum-ru-issues-and-discussions#162] [Как это сделать?] [answered]
Подскажите, как реализовать генератор qr кодов внутри тотум? Ткните что прочитать.  
менеджеры печатают бланки из интерфейса тотум, для клиентов я сделал некое API ( кривое косое ) где можно в онлайне смотреть свои заказы. Было бы очень круто печатать не просто ссылку на сервис, а динамически генерировать для каждого номера свой qr и выводить в печать.

**Comment by erpic-ru** [comment, marked as answer]
Я делал так. Берете любой бесплатный онлайн генератор кодов с API. Через getfromscript отправляете туда нужные вам ссылку или текст, в ответ получаете ссылку на изображение qr-кода (подберете по формату) и вставляете ее в печатный шаблон `<img src="URL" >
`

**Reply by Qb1cc** [reply]
[https://ru.docs.totum.online/calling-a-third-party-script-from-button-code](https://ru.docs.totum.online/calling-a-third-party-script-from-button-code) разъяснения по getfromscript , полные.  Сначал не мог найти как ответ записать. И еще я так понимаю что использовать это не совсем безопасно с  бесплатными сервисами, т.к таблица может зависнуть если сервис от которого ждём ответ, зависает. Попробую написать свой генератор, если заведу, скину скрипт.

**Reply by ttm-alexeypolunin** [reply]
Можно асинхронно делать эти обращения к стороннему сервису — записываете сначала в Тотум в отдельную таблицу строку, а потом по галочке из этой спецтаблицы запрашиваете сторонний сервис. Галочки ставите кроном.

Мы планировали генератор втроенный в PRO, но походу он будет и в MIT-версии.

Также если вы найдете генератор, который можно положить на сервере, то его можно вызывать через `execSSH`.

И еще можно `getFromScript` скормить путь к картинке и добавить ее сохранить ее себе в Тотум. Где-то к полю `Файл` есть инструкция.

**Reply by ttm-alexeypolunin** [reply]
Для тех, кто будет смотреть потом небольшой пример:

```
= : linkToDataHtml(title: "Qr code"; html: $replace)

replace: strReplace(str: $html; from: json`["{DATA}","{SIZE}"]`; to: json`[$data,$size]`)

    ```html:html
    <img src="https://api.qrserver.com/v1/create-qr-code/?data={DATA}&size={SIZE}"/>
    ```

data: "https://totum.online"
size: "300x300"
```

Вот документация: https://goqr.me/api/doc/create-qr-code/

<img width="663" alt="Снимок экрана 2022-06-24 в 12 37 44" src="https://user-images.githubusercontent.com/55755565/175508671-d1c12f56-5761-45d8-8c1a-e943f24edd50.png">

**Comment by ivgo** [comment]
Доброго времени суток, Может быть кому пригодится. Я сам не использовал API так как не являюсь программистом а простой админ эникейщик. Но делал вывод  через формы для печати.
Для простого штрихкода просто подключил шрифты с гуугла.А для QR  через POST GET и подтягивание картинки.
`<link href='https://fonts.googleapis.com/css?family=Libre Barcode 128' rel='stylesheet'>
 <style type="text/css">
   h2 {font-family: 'Libre Barcode 128';font-size: 42px;}`


Там где нужен штрих код просто оформяю как заголовок 2
